Core/clim-basic/utils.lisp

```diff
@@ -743,3 +728,195 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
+                                                pos
+                                                (return-from breaks-rec nil))))
+                           (multiple-value-bind (breaks remaining-cost)
+                               (breaks-rec 0 margin current-break end (1+ i))
+                             (when (or (null best-cost)
+                                       (< (+ current-cost remaining-cost) best-cost))
+                               (setf best-breaks (list* current-break breaks)
+                                     best-cost (+ current-cost remaining-cost))))))
+                        ((null best-cost)
+                         (return-from breaks-rec
+                           (if (> offset 0)
+                               (multiple-value-bind (breaks remaining-cost)
+                                   (breaks-rec 0 margin start end i)
+                                 (return-from breaks-rec
+                                   (values (list* start breaks)
+                                           (+ remaining-cost current-margin))))
+                               (let* ((char-breaks (%line-breaks-1 string width
+                                                                   offset margin
+                                                                   start opportunity))
+                                      (new-start (alexandria:last-elt char-breaks)))
+                                 (multiple-value-bind (breaks remaining-cost)
+                                     (breaks-rec 0 margin new-start end i)
+                                   (values (append char-breaks breaks) remaining-cost))))))
+                        (T #1=(return-from breaks-rec
+                                (values best-breaks best-cost))))
+                finally #1#)))
+    (breaks-rec initial-offset margin start end 0)))
+
+;; Implementing line breaking as defined in Unicode[1] is left as an excercise
+;; for the reader. [1] https://unicode.org/reports/tr14/ -- jd 2019-01-08
+(defun line-breaks (string width &key (break-strategy t) initial-offset margin (start 0) end)
+  "Function takes a string and returns a list of indexes where it should be split.
+
+WIDTH is a function accepting STRING, START and END arguments which should
+return string width for these boundaries. Alternatively for fixed font width it
+is a number.
+
+INITIAL-OFFSET is an initial position for the first line (may be negative). All
+remaining lines will start from the line beginning. Default is line beginning.
+
+MARGIN is a maximum width at which line should break. Defaults to 80
+characters (width of a character m is taken as a reference value).
+
+BREAK-STRATEGY may be:
+- symbol T implementing a default line breaking by word strategy,
+- symbol NIL implementing a line breaking by character strategy,
+- function accepting index which should return T for break opportunity,
```

Comment:
Could `t` and `nil` be replaced with something like `:at-word-boundaries` and `:anywhere`?